### PR TITLE
fix(grpo): add opt-in mask_zero_advantage_groups for zero-variance groups under beta>0

### LINKED ISSUE #5588: [Bug] Zero-std reward groups produce spurious KL gradients when `beta > 0` in GRPO/RLOO
### Reproduction

When all completions in a group receive the same reward, advantages are zero:
```python
advantages = rewards - mean_grouped_rewards  # = 0 for all completions
```

So the policy loss contribution is zero, no learning signal, as expected. However, the KL penalty term is **not** zero:
```python
per_token_loss = per_token_loss + self.beta * per_token_kl  # per_token_kl ≠ 0
```

`per_token_kl` has gradients through `per_token_logps` (the current model's log-probs), while `ref_per_token_logps` is detached (computed under `torch.no_grad()`). So zero-std groups still generate gradients that push the model toward the reference policy, even though there's no reward signal guiding the direction. KL regularization should only constrain the model when there's actual reward signal to balance against. Without it, KL gradients from zero-std groups act as undirected noise.

This affects both `GRPOTrainer` and `RLOOTrainer` (RLOO defaults to `beta=0.05`).

A simple fix is to zero out `completion_mask` for zero-std groups, which suppresses both the (already-zero) policy loss and the spurious KL gradients:
```python
is_std_zero_local = is_std_zero[process_slice]
completion_mask = completion_mask * (~is_std_zero_local).unsqueeze(1).int()
```

This is orthogonal to any future oversampling/dynamic sampling implementation, even if oversampling is added to backfill zero-std groups with new prompts, any groups that exhaust resampling attempts and still have zero std should be masked for the same reason.

### Reproducing Code
```python
import torch
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

def constant_reward(completions, **kwargs):
    return [1.0] * len(completions)

dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")

for beta in [0.0, 0.04]:
    trainer = GRPOTrainer(
        model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
        reward_funcs=constant_reward,
        args=GRPOConfig(
            output_dir=f"/tmp/kl_bug_{beta}",
            per_device_train_batch_size=3,
            num_generations=3,
            max_completion_length=8,
            beta=beta,
            max_steps=3,
            report_to="none",
        ),
        train_dataset=dataset,
    )

    params_before = {n: p.clone() for n, p in trainer.model.named_parameters()}
    trainer.train()
    changed = sum(
        not torch.equal(p, trainer.model.get_parameter(n))
        for n, p in params_before.items()
    )
    print(f"\nbeta={beta}: {changed}/{len(params_before)} params changed")
    # beta=0.0 -> 0 params changed (correct: no reward signal, no update)
    # beta=0.04 -> ALL params changed (bug: KL gradients update model with no reward signal)
```

### output
```
beta=0.0: 0/27 params changed
beta=0.04: 27/27 params changed
```

### System Info

```
- Platform: Linux-6.8.0-1043-nvidia-x86_64-with-glibc2.39
- Python version: 3.12.13
- TRL version: 1.2.0.dev0+573ea22
- PyTorch version: 2.11.0+cu130
- accelerator(s): NVIDIA B200, NVIDIA B200, NVIDIA B200, NVIDIA B200, NVIDIA B200, NVIDIA B200, NVIDIA B200, NVIDIA B200
- Transformers version: 5.5.3
- Accelerate version: 1.13.0
- Accelerate config: 
  - compute_environment: LOCAL_MACHINE
  - distributed_type: DEEPSPEED
  - mixed_precision: bf16
  - use_cpu: False
  - debug: False
  - num_processes: 8
  - machine_rank: 0
  - num_machines: 1
  - rdzv_backend: static
  - same_network: True
  - main_training_function: main
  - enable_cpu_affinity: False
  - deepspeed_config: {'gradient_accumulation_steps': 1, 'gradient_clipping': 1.0, 'offload_optimizer_device': 'none', 'offload_param_device': 'none', 'zero3_init_flag': True, 'zero_stage': 2}
  - downcast_bf16: no
  - tpu_use_cluster: False
  - tpu_use_sudo: False
  - tpu_env: []
- Datasets version: 4.8.4
- HF Hub version: 1.11.0
- bitsandbytes version: not installed
- DeepSpeed version: 0.18.9
- Liger-Kernel version: 0.7.0
- PEFT version: not installed
- vLLM version: dev1+gba4a78eb5.cu131
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by karthikchundi-commits**
# What does this PR do?

Fixes # (issue)

Follow-up to #5588 ("Zero-std reward groups produce spurious KL gradients when `beta > 0`"), which was closed by @qgallouedec as expected GRPO behavior rather than a bug — with a suggestion that if there's interest, it could be added as an opt-in flag (e.g. `mask_zero_advantage_groups: bool = False`) referencing the paper raised in that discussion. This PR implements exactly that.

**What it does:** when `beta > 0` and every completion in a reward group gets the same reward, the group's advantage is exactly zero, so the policy loss term drops out — but the per-token KL penalty still applies regardless of the advantage, so the group still contributes a gradient. `GRPOConfig.mask_zero_advantage_groups` (default `False`, per `CONTRIBUTING.md`'s opt-in-for-new-features guideline) masks the completion mask for these zero-std groups when enabled, matching the zero-variance masking proposed in [Revisiting Group Relative Policy Optimization](https://huggingface.co/papers/2505.22257), Section 3.2.

Also checked whether RLOO needs the same flag (the original issue's title mentioned both trainers): it doesn't. RLOO folds the KL penalty into the reward *before* computing `std_rewards`/advantages, so a genuinely zero-std group there already has zero gradient by construction — there's no separate KL loss term the way GRPO's `compute_loss` adds one. Left RLOO untouched rather than adding a no-op flag.

Updated `docs/source/paper_index.md` per `.ai/AGENTS.md`'s requirement for PRs implementing a paper's method, and added a matching `docs/source/grpo_trainer.md` tip next to the existing `beta` documentation.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? #5588, see maintainer comment: https://github.com/huggingface/trl/issues/5588#issuecomment-3282898242
- [x] Did you make sure to update the documentation with your changes? `docs/source/grpo_trainer.md`, `docs/source/paper_index.md`
- [x] Did you write any new necessary tests? `tests/test_grpo_trainer.py::TestGRPOTrainer::test_train_with_mask_zero_advantage_groups`

## AI writing disclosure

- [x] AI-generated: the PR was mostly or fully generated by an AI tool (Claude Code), under my direction and review. I understand the change, engaged with the maintainer's original issue discussion in scoping it, and can discuss/iterate on it.

## Who can review?

@qgallouedec — tagging since you closed #5588 and suggested this exact opt-in approach.

---

> [!NOTE]
> **Low Risk**
> Opt-in config defaulting to False; only affects loss masking when enabled, mainly relevant for KL-regularized GRPO runs.
> 
> **Overview**
> Adds **`GRPOConfig.mask_zero_advantage_groups`** (default `False`) so training can match zero-variance masking from [Revisiting GRPO](https://huggingface.co/papers/2505.22257) when **`beta > 0`**.
> 
> With identical rewards in a group, advantages are zero so the policy term vanishes, but the per-token KL penalty can still update weights. When the flag is enabled, **`GRPOTrainer`** zeros **`completion_mask`** for those zero–reward-std groups before loss, removing both policy and KL contributions from them. Default behavior is unchanged.
> 
> Docs add a tip in **`grpo_trainer.md`**, a **`paper_index.md`** entry, and **`test_train_with_mask_zero_advantage_groups`** checks that parameters move only from KL when masking is off and stay fixed when it is on.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b4ff0c8a0076a61d7d7b6ad58b50d3813884f577. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Closing: we don't review fully AI-generated PRs from first-time contributors, see the [AI usage policy](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#ai-usage-policy).
